airbyte-workers/src/main/java/io/airbyte/workers/temporal/TemporalClient.java

```diff
@@ -446,6 +451,79 @@ public ManualOperationResult synchronousResetConnection(final UUID connectionId,
         Optional.of(resetJobId), Optional.empty());
   }
 
+  public void restartWorkflowByStatus(final WorkflowExecutionStatus executionStatus) {
+    final Set<String> workflowExecutionInfos = fetchWorkflowByStatus(executionStatus);
+
+    final Set<String> nonRunningWorkflow = filterOutRunningWorkspaceId(workflowExecutionInfos);
+
+    nonRunningWorkflow.forEach(workflowId -> {
+      final Optional<UUID> maybeConnectionId = extractConnectionIdFromWorkflowId(workflowId);
+      log.error(workflowId);
+      if (maybeConnectionId.isPresent()) {
+        ConnectionManagerUtils.safeTerminateWorkflow(client, maybeConnectionId.get(), "Terminating workflow in "
+            + "unreachable state before starting a new workflow for this connection");
+        ConnectionManagerUtils.startConnectionManagerNoSignal(client, maybeConnectionId.get());
+      } else {
+        log.error("Malformated workflow id, won't terminate of restart");
+      }
+    });
+  }
+
+  /**
+   * This should be in the class {@li}
+   * @param workflowId
+   * @return
+   */
+  Optional<UUID> extractConnectionIdFromWorkflowId(final String workflowId) {
+    if (!workflowId.startsWith("connection_manager_")) {
+      return Optional.empty();
+    }
+    System.err.println(workflowId);
+    System.err.println(StringUtils.removeStart(workflowId, "connection_manager_"));
+    System.err.flush();
+    return Optional.ofNullable(StringUtils.removeStart(workflowId, "connection_manager_"))
+        .map(
+            stringUUID ->
+                UUID.fromString(stringUUID));
+  }
+
+  Set<String> fetchWorkflowByStatus(final WorkflowExecutionStatus executionStatus) {
```

Comment:
```suggestion
  Set<String> fetchWorkflowsByStatus(final WorkflowExecutionStatus executionStatus) {
```

This is a plural method, that returns a set